audit.go

```diff
@@ -201,21 +211,42 @@ func main() {
 	//Main loop. Get data from netlink and send it to the json lib for processing
 	for {
 		msg, err := nlClient.Receive()
-		timing := metric.GetClient().NewTiming() // measure latency from recipt of message
 		if err != nil {
 			if err.Error() == "no buffer space available" {
 				metric.GetClient().Increment("messages.netlink_dropped")
 			}
 			slog.Error.Printf("Error during message receive: %+v\n", err)
 			continue
 		}
-
-		metric.GetClient().Increment("messages.total")
 		if msg == nil {
 			continue
 		}
+		// As soon as we have a message, spawn a goroutine to handle it and free up the main loop
+		go handleMsg(msg, marshaller)
```

Comment:
This needs to have recovery in case of a panic.

Reply:
Handled in 13c19f19d975057eff14aa50bdb6b4f4c67f7adf

Reply:
At a very high level this makes sense. I note you do acknowledge that this could spawn more go routines than resources can support with an unbounded number of go routines. Do we know how this would manifest itself?

Reply:
I think itd go like:
- main loop processing as fast as it can, kicking off goroutines
  - this spawns goroutines faster than they finish
    - each routine also creates its own buffer to read the messages vs. a shared one that was in use before
  - each new goroutine eats up cpu/memory until the pod OOMs

IDK if it'd actually be an issue, but something that could happen. Maybe we cap the processing time of a goroutine and kill it if > 10s or something?

Reply:
Gotcha, just wanted to capture what to expect if/when it happens. Not sure how difficult it would be to have workers/channels to "regulate" the number of go routines. Thanks for the explanation though.

Reply:
totally untested code, but something like this?
```go
    msgChan := make(chan *syscall.NetlinkMessage, 100) // Buffer size can be adjusted

    // Start worker pool
    numWorkers := 10
    var wg sync.WaitGroup
    for i := 0; i < numWorkers; i++ {
        wg.Add(1)
        go worker(msgChan, marshaller, &wg)
    }

    for {
        msg, err := nlClient.Receive()
        if err != nil {
            if err.Error() == "no buffer space available" {
                metric.GetClient().Increment("messages.netlink_dropped")
            }
            slog.Error.Printf("Error during message receive: %+v\n", err)
            continue
        }
        if msg == nil {
            continue
        }
        msgChan <- msg
    }

    close(msgChan)
    wg.Wait()
```

with a worker like:

```go
    func worker(msgChan <-chan *syscall.NetlinkMessage, marshaller *marshaller.AuditMarshaller, wg *sync.WaitGroup) {
    defer wg.Done()
    for msg := range msgChan {
        handleMsg(msg, marshaller)
    }
}
```